### ISSUE TITLE
Some questions about the training code

### ISSUE THREAD

**Issue by Tonsty** [lpiccinelli-eth/UniDepth#86]
### **Question 1 - the padding may be assigned in incorrect  order **

During ContextCrop process, the padding is assigned as follows
https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/pipelines/transforms.py#L1275-L1280

which seem to be in the order of  [left, bottom, right, top], not match the following code 

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/pipelines/transforms.py#L1307

but I do not determine whether it matters because I found the transform operations are performed use following code

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/pipelines/transforms.py#L1315-L1320

### **Question 2 - full contained condition may not be satisfied **

In the following code, the second branch condition _"output_ratio / input_ratio * ctx > 1"_ will never meet because _output_ratio / input_ratio <= 1.0_ and _ctx < 1.0_.  And in the "fully contained" branch, "full contained" condition may either not be satisfied because 
_new_h = new_w / output_ratio
           =  w * (ctx * output_ratio / input_ratio) ** 0.5 / output_ratio
           =  h * input_radio * (ctx * output_ratio / input_ratio) ** 0.5 / output_ratio
           = h * (ctx * input_radio / out_radio) ** 0.5_
although _ctx < 1.0_, _input_radio / out_radio_ is larger than 1.0, so that _new_h_ may be larger than _h_, then the crop cannot be fully contained.

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/pipelines/transforms.py#L1213-L1225

### **Question 3 -  Only half of the batched data are used for training?**

_nsteps_accumulation_gradient_ is set to 1 in the training config, and _batch_chunk_ is equal to _batch_size_, but the batched data obtained from dataloader (_batches["data"]_) in fact have a double batch size of _2 * batch_size_ since a pair of data items are loaded in _ConcatDataset.\_\_get_item\_\__  (_self.pairs_ is configured to 2) . Thus, only the first half data in _batches["data"]_ are used for training.

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/scripts/train.py#L258

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/scripts/train.py#L423-L430

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/utils.py#L49-L55

### **Question 4 -  SequenceDataset is not suitable for training because of the calculation of SelfDistill loss **

The SelfDistill loss is under the assumption that each pair of image from the _chunks_ is at the same view, which in fact originates from the same image but with different crop/resize operations. However,  given certain idx, _super(ConcatDataset, self).\_\_getitem\_\_(idx)_ may load two different images from the underlying image sequence (_start_ is a random value between 0 and _num_samples_sequence_).

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/ops/losses/distill.py#L27

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/utils.py#L49-L55

https://github.com/lpiccinelli-eth/UniDepth/blob/5afc0dc7a1d714ecc656b218982b0aa6ef5e8585/unidepth/datasets/sequence_dataset.py#L219-L224

**Comment by lpiccinelli-eth** [comment]
Q1: you are right that they are different, but because the evaluation and data processing use different padding torchvision and original torch. So in the dataloading case is [left,top,right,bottom] while for eval is [left,right,top,bottom]. In the original (private) repo we have a method of the UniDepth class that rearranges the padding.

Q2: the case you are referring to is when one dimension of the image is padded and the other is cropped, and you are right, the inequality has to be flipped. However, that kind of case can still happen, but the new image shape is calculated differently from what we expect based on `ctx`.

Q3 and Q4 have been solved with the latest commit. The original repo dataloader and the public one were different and caused these issues.
